### PR TITLE
[WIP] support for pytorch mps (metal/macos)

### PR DESCRIPTION
**Pull request by tmm1**
with these PR, (once I uncomment pytorch-mps entries in `Pipfile`), I can run `make setup`, `make wave` and `train.py` successfully on macOS.

```
$ python3.10 -m pipenv run python train.py -C examples/cfg_example_oasst1.py
                                                                                                                           
===================================BUG REPORT===================================                                           
Welcome to bitsandbytes. For bug reports, please run                                                                       
                                                                                                                           
python -m bitsandbytes                                                                                                     
                                                                                                                           
 and submit this information together with your error trace to: https://github.com/TimDettmers/bitsandbytes/issues         
================================================================================                                           
bin /Users/tmm1/.local/share/virtualenvs/h2o-llmstudio-kx1PJ3DV/lib/python3.10/site-packages/bitsandbytes/libbitsandbytes_c
pu.so                                                                                                                      
/Users/tmm1/.local/share/virtualenvs/h2o-llmstudio-kx1PJ3DV/lib/python3.10/site-packages/bitsandbytes/cextension.py:34: Use
rWarning: The installed version of bitsandbytes was compiled without GPU support. 8-bit optimizers, 8-bit multiplication, a
nd GPU quantization are unavailable.                                                                                       
  warn("The installed version of bitsandbytes was compiled without GPU support. "                                          
'NoneType' object has no attribute 'cadam32bit_grad_fp32'                                                                  
CUDA SETUP: Loading binary /Users/tmm1/.local/share/virtualenvs/h2o-llmstudio-kx1PJ3DV/lib/python3.10/site-packages/bitsand
bytes/libbitsandbytes_cpu.so...                                                                                            
dlopen(/Users/tmm1/.local/share/virtualenvs/h2o-llmstudio-kx1PJ3DV/lib/python3.10/site-packages/bitsandbytes/libbitsandbyte
s_cpu.so, 0x0006): tried: '/Users/tmm1/.local/share/virtualenvs/h2o-llmstudio-kx1PJ3DV/lib/python3.10/site-packages/bitsand
bytes/libbitsandbytes_cpu.so' (not a mach-o file), '/System/Volumes/Preboot/Cryptexes/OS/Users/tmm1/.local/share/virtualenv
s/h2o-llmstudio-kx1PJ3DV/lib/python3.10/site-packages/bitsandbytes/libbitsandbytes_cpu.so' (no such file), '/Users/tmm1/.lo
cal/share/virtualenvs/h2o-llmstudio-kx1PJ3DV/lib/python3.10/site-packages/bitsandbytes/libbitsandbytes_cpu.so' (not a mach-
o file)                                                                                                                    
2023-06-30 13:30:45,198 - INFO: Global random seed: 1                                                                      
2023-06-30 13:30:45,199 - INFO: Preparing the data...                                                                      
2023-06-30 13:30:45,199 - INFO: Setting up automatic validation split...                                                   
2023-06-30 13:30:45,286 - INFO: Preparing train and validation data                                                        
2023-06-30 13:30:45,286 - INFO: Loading train dataset...                                                                   
Using pad_token, but it is not set yet.                                                                                    
Using cls_token, but it is not set yet.                                                                                    
Using sep_token, but it is not set yet.                                                                                    
2023-06-30 13:30:45,735 - INFO: Stop token ids: []                                                                         
2023-06-30 13:30:45,763 - INFO: Sample prompt: Can you make it about 50% shorter and more exciting? Think to a dynamic tv a
dvertisement for sport product<|endoftext|>                                                                                
2023-06-30 13:30:45,763 - INFO: Loading validation dataset...                                                              
Using pad_token, but it is not set yet.                                                                                    
Using cls_token, but it is not set yet.                                                                                    
Using sep_token, but it is not set yet.                                     
2023-06-30 13:30:45,735 - INFO: Stop token ids: []                                                                         
2023-06-30 13:30:45,763 - INFO: Sample prompt: Can you make it about 50% shorter and more exciting? Think to a dynamic tv a
dvertisement for sport product<|endoftext|>                                                                                
2023-06-30 13:30:45,763 - INFO: Loading validation dataset...                                                              
Using pad_token, but it is not set yet.                                                                                    
Using cls_token, but it is not set yet.                                                                                    
Using sep_token, but it is not set yet.                                                                                    
2023-06-30 13:30:46,060 - INFO: Stop token ids: []                                                                         
2023-06-30 13:30:46,061 - INFO: Sample prompt: What types of tests do we have in software development?<|endoftext|>        
2023-06-30 13:30:46,061 - INFO: Number of observations in train dataset: 8273                                              
2023-06-30 13:30:46,061 - INFO: Number of observations in validation dataset: 1                                            
2023-06-30 13:30:46,163 - INFO: Using float32 for backbone                                                                 
trainable params: 131,072 || all params: 1,011,912,704 || trainable%: 0.012952895984197467                                 
2023-06-30 13:30:48,190 - INFO: Enough space available for saving model weights.                                           
2023-06-30 13:30:48,255 - INFO: Starting validation inference                                                              
2023-06-30 13:30:48,255 - INFO: validation progress:   0%|          | 0/1 [00:00<?, ?it/s]                                 
2023-06-30 13:30:58,759 - INFO: validation progress: 100%|##########| 1/1 [00:10<00:00, 10.50s/it]                         
2023-06-30 13:30:58,762 - INFO: validation progress: 100%|##########| 1/1 [00:10<00:00, 10.51s/it]
2023-06-30 13:30:58,770 - INFO: Mean validation loss: 1.90946 
2023-06-30 13:30:58,773 - INFO: Validation BLEU: 0.01722
2023-06-30 13:30:58,795 - INFO: Training Epoch: 1 / 1
2023-06-30 13:30:58,795 - INFO: train loss:   0%|          | 0/2068 [00:00<?, ?it/s]
Using pad_token, but it is not set yet.
Using cls_token, but it is not set yet.
Using sep_token, but it is not set yet.
2023-06-30 13:30:59,041 - INFO: Stop token ids: []
/Users/tmm1/.local/share/virtualenvs/h2o-llmstudio-kx1PJ3DV/lib/python3.10/site-packages/torch/utils/checkpoint.py:418: Use
rWarning: torch.utils.checkpoint: please pass in use_reentrant=True or use_reentrant=False explicitly. The default value of
 use_reentrant will be updated to be False in the future. To maintain current behavior, pass use_reentrant=True. It is reco
mmended that you use use_reentrant=False. Refer to docs for more details on the differences between the two variants.
  warnings.warn(
2023-06-30 13:32:09,685 - INFO: train loss: 1.36:   5%|4         | 103/2068 [01:10<22:32,  1.45it/s]
2023-06-30 13:32:28,336 - INFO: train loss: 1.36:   5%|4         | 103/2068 [01:29<22:32,  1.45it/s]
...
2023-06-30 13:48:54,841 - INFO: train loss: 1.17: 100%|##########| 2068/2068 [17:56<00:00,  1.92it/s]
2023-06-30 13:48:54,841 - INFO: Starting validation inference
2023-06-30 13:48:54,841 - INFO: validation progress:   0%|          | 0/1 [00:00<?, ?it/s]
2023-06-30 13:49:00,265 - INFO: validation progress: 100%|##########| 1/1 [00:05<00:00,  5.42s/it]
2023-06-30 13:49:00,267 - INFO: validation progress: 100%|##########| 1/1 [00:05<00:00,  5.43s/it]
2023-06-30 13:49:00,275 - INFO: Mean validation loss: 0.86673
2023-06-30 13:49:00,277 - INFO: Validation BLEU: 0.49945
2023-06-30 13:49:00,317 - INFO: Saving last model checkpoint: val_loss 0.86673, val_BLEU 0.49945 to examples/output_oasst1/
[1]    37069 segmentation fault  python3.10 -m pipenv run python train.py -C examples/cfg_example_oasst1.py
```

### CODE REVIEW

**Comment by tmm1** [comment]
> uncomment pytorch-mps entries in `Pipfile`

unfortunately `pipenv` doesn't support multi-platform selection: https://github.com/pypa/pipenv/issues/5130

i tried to implement `pipenv` categories, with a separate `[mps-packages]` section in the `Pipfile`- but that didn't work either

for now i've just left things commented out, but one viable approach may be a `./torch-select` shim package: https://github.com/pypa/pipenv/issues/2171#issuecomment-387877895

File: examples/cfg_example_oasst1.py [hunk 1]
```diff
@@ -49,7 +50,7 @@ class Config(ConfigProblemBase):
         token_mask_probability=0.0
     )
     architecture: ConfigNLPCausalLMArchitecture = ConfigNLPCausalLMArchitecture(
-        backbone_dtype="float16",
+        backbone_dtype="float32" if torch.backends.mps.is_available() else "float16",
```
Comment (tmm1) [inline examples/cfg_example_oasst1.py:53]: mps doesn't support float16 currently, but support is planned and Apple announced macOS Sonoma released this fall will support AMP

https://github.com/pytorch/pytorch/issues/88415
https://github.com/pytorch/pytorch/issues/104191

File: train.py [hunk 2]
```diff
@@ -65,6 +65,20 @@
...
+    multiprocessing.set_start_method("fork")
+
+    # workaround missing autocast support in pytorch
+    # https://github.com/pytorch/pytorch/issues/104478
+    from types import SimpleNamespace
+    torch.mps.amp = SimpleNamespace()
+    torch.mps.amp.autocast = autocast
+    torch.mps._initialized = False
+    torch.mps.get_autocast_dtype = lambda: torch.float32
+    torch.mps.is_autocast_enabled = lambda: False
```
Comment (tmm1) [inline train.py:81]: cc https://github.com/pytorch/pytorch/issues/104478

**Comment by psinger** [comment]
Hi @tmm1 - you could consider making a separate pipenv for mac installs, although not ideal. Would be better to figure out a single-stop solution. Any new insights on this?

File: examples/cfg_example_oasst1.py (hunk 1 above)
Comment (pascal-pfeiffer) [reply examples/cfg_example_oasst1.py:53]: In this case, I would hide the setting on macOS using the `_visibility` flag and additionally replace any settings != float32 with a warning for CLI users

**Comment by pascal-pfeiffer** [comment]
Thank you for working on this @tmm1. Did you find any solution to the issue yet? 
Another option would be to switch to poetry maybe, which should support multi-platform. I only know from experience that there have been some issues with pytorch versions in the past, but the current version seems to handle that well.

**Comment by USMCM1A1** [comment]
@tmm1 I'm really interested in running H2O LLM Studio on an a new M2 max--have you been able to make any progress?

**Comment by psinger** [comment]
Closing this for now - please reopen if there is interest following up on it. Thanks!

**Comment by qdrddr** [comment]
Maybe it'll be usefull
https://apeatling.com/articles/part-3-fine-tuning-your-llm-using-the-mlx-framework/

**Pull request closed without merging**